File: java-core-plugins/src/main/java/org/unlaxer/tramli/plugins/eventstore/EventLogStoreDecorator.java
package org.unlaxer.tramli.plugins.eventstore;

import org.unlaxer.tramli.FlowContext;
import org.unlaxer.tramli.FlowDefinition;
import org.unlaxer.tramli.FlowInstance;
import org.unlaxer.tramli.FlowState;
import org.unlaxer.tramli.FlowStore;

import java.time.Instant;
import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;

public final class EventLogStoreDecorator implements FlowStore {
    private final FlowStore delegate;
    private final List<VersionedTransitionEvent> events = new ArrayList<>();
    private final Map<String, Long> versions = new LinkedHashMap<>();

    public EventLogStoreDecorator(FlowStore delegate) {
        this.delegate = delegate;
    }

    @Override
    public void create(FlowInstance<?> flow) {
        delegate.create(flow);
        versions.put(flow.id(), 0L);
    }

    @Override
    public <S extends Enum<S> & FlowState> Optional<FlowInstance<S>> loadForUpdate(String flowId, FlowDefinition<S> definition) {
        return delegate.loadForUpdate(flowId, definition);
    }

    @Override
    public void save(FlowInstance<?> flow) {
        delegate.save(flow);
    }

    @Override
    public void recordTransition(String flowId, FlowState from, FlowState to, String trigger, FlowContext ctx) {
        long nextVersion = nextVersion(flowId);
        events.add(new VersionedTransitionEvent(
                nextVersion,
                flowId,
                from != null ? from.name() : null,
                to != null ? to.name() : null,
                trigger,
                Instant.now(),
                snapshot(ctx),
                "TRANSITION",
                Map.of()));
        delegate.recordTransition(flowId, from, to, trigger, ctx);
    }

    public void recordCompensation(String flowId, String compensationName, Map<String, Object> metadata) {
        long nextVersion = nextVersion(flowId);
        events.add(new VersionedTransitionEvent(
                nextVersion,
                flowId,
                null,
                null,
                compensationName,
                Instant.now(),
                Map.of(),
                "COMPENSATION",
                Map.copyOf(metadata)));
    }

    public List<VersionedTransitionEvent> events() {
        return Collections.unmodifiableList(events);
    }

    public List<VersionedTransitionEvent> eventsForFlow(String flowId) {
        List<VersionedTransitionEvent> result = new ArrayList<>();
        for (VersionedTransitionEvent event : events) {
            if (flowId.equals(event.flowId())) {
                result.add(event);
            }
        }
        return Collections.unmodifiableList(result);
    }

    private long nextVersion(String flowId) {
        return versions.compute(flowId, (k, v) -> v == null ? 1L : v + 1L);
    }

    private Map<String, Object> snapshot(FlowContext ctx) {
        Map<String, Object> snapshot = new LinkedHashMap<>();
        for (Map.Entry<Class<?>, Object> e : ctx.snapshot().entrySet()) {
            snapshot.put(e.getKey().getSimpleName(), e.getValue());
        }
        return snapshot;
    }
}

File: ts-plugins/src/eventstore/event-log-store-decorator.ts
import type { FlowContext } from '@unlaxer/tramli';
import type { VersionedTransitionEvent } from './types.js';

/**
 * FlowStore decorator that maintains an append-only event log.
 * Tenure-lite: not full event sourcing, intentionally lighter.
 */
export class EventLogStoreDecorator {
  private eventLog: VersionedTransitionEvent[] = [];
  private versionCounters = new Map<string, number>();

  constructor(private readonly delegate: any) {}

  create(flow: any): void { this.delegate.create(flow); }

  loadForUpdate<S extends string>(flowId: string, definition?: any): any {
    return this.delegate.loadForUpdate(flowId, definition);
  }

  save(flow: any): void { this.delegate.save(flow); }

  recordTransition(flowId: string, from: any, to: string, trigger: string, ctx: FlowContext): void {
    this.delegate.recordTransition(flowId, from, to, trigger, ctx);
    const version = (this.versionCounters.get(flowId) ?? 0) + 1;
    this.versionCounters.set(flowId, version);
    this.eventLog.push({
      flowId, version, type: 'TRANSITION',
      from: from?.toString() ?? null, to, trigger,
      timestamp: new Date(),
      stateSnapshot: JSON.stringify(Object.fromEntries(ctx.snapshot())),
    });
  }

  /** All events across all flows. */
  events(): readonly VersionedTransitionEvent[] { return this.eventLog; }

  /** Events for a specific flow. */
  eventsForFlow(flowId: string): VersionedTransitionEvent[] {
    return this.eventLog.filter(e => e.flowId === flowId);
  }

  /** Append a compensation event. */
  appendCompensation(flowId: string, trigger: string, metadata: string): void {
    const version = (this.versionCounters.get(flowId) ?? 0) + 1;
    this.versionCounters.set(flowId, version);
    this.eventLog.push({
      flowId, version, type: 'COMPENSATION',
      from: null, to: 'COMPENSATED', trigger,
      timestamp: new Date(), stateSnapshot: metadata,
    });
  }

  get transitionLog() { return this.delegate.transitionLog; }
  clear(): void { this.delegate.clear?.(); this.eventLog = []; this.versionCounters.clear(); }
}
